lib/sycamore/sycamore/writers/base.py

```diff
@@ -0,0 +1,160 @@
+    class record_t(ABC):
+
+        @classmethod
+        @abstractmethod
+        def from_doc(cls, document: Document) -> "BaseDBWriter.record_t":
+            pass
+
+        @abstractmethod
+        def serialize(self) -> bytes:
+            pass
+
+        @classmethod
+        @abstractmethod
+        def deserialize(cls, byteses: bytes) -> "BaseDBWriter.record_t":
+            pass
+
+    # Type param for the object used to configure a new index if necessary
+    @dataclass
+    class index_params_t(ABC):
+        pass
+
+    # Type param for the object used to create a client
+    @dataclass
+    class client_params_t(ABC):
+        pass
+
+    def __init__(
+        self, plan: Node, client_params: client_params_t, index_params: Optional[index_params_t], **ray_remote_args
+    ):
+        super().__init__(plan, **ray_remote_args)
+        _check_serializable(client_params, index_params)
+
+        self._client_params = client_params
+        self._index_params = index_params
+
+    def get_client(self, client_params: client_params_t) -> client_t:
+        return self.client_t.from_client_params(client_params)
+
+    @classmethod
+    def doc_to_record(cls, doc: Document) -> record_t:
+        return cls.record_t.from_doc(doc)
+
+    def execute(self) -> Dataset:
+        input_dataset = self.child().execute()
+        record_dataset = self._ray_map_docs_to_records(input_dataset)
+        self.get_client(self._client_params)
+        datasink = BaseDBWriter.InnerDatasink(self._client_params, self._index_params, self.__class__)
+        _check_serializable(datasink)
+        record_dataset.write_datasink(datasink, ray_remote_args=self.resource_args)
+        return input_dataset
+
+    def _ray_map_docs_to_records(self, dataset: Dataset) -> Dataset:
+
+        def build_ray_callable():
+            cls = self.__class__
+
+            def ray_callable(ray_input: dict[str, np.ndarray]) -> dict[str, list]:
+                all_docs = [Document.deserialize(s) for s in ray_input.get("doc", [])]
+                docs = [d for d in all_docs if not isinstance(d, MetadataDocument)]
+                records = [cls.doc_to_record(d) for d in docs]
+                return {"record": [r.serialize() for r in records]}
+
+            return ray_callable
+
+        ray_callable = build_ray_callable()
+        _check_serializable(ray_callable)
+
+        return dataset.map_batches(ray_callable, **self.resource_args)
+
+    class InnerDatasink(Datasink):
+
+        def __init__(
+            self,
+            client_params: "BaseDBWriter.client_params_t",
+            index_params: Optional["BaseDBWriter.index_params_t"],
+            owner_cls: type["BaseDBWriter"],
+        ):
+            _check_serializable(client_params, index_params, owner_cls)
+
+            self._client_params = client_params
+            self._index_params = index_params
+            self._owner = owner_cls
+
+        def on_write_start(self) -> None:
```

Comment:
This is one piece that I'm not sure how best to handle. Maybe if we define it as create_index_idempotent it will work, and the logic becomes create if missing -> check if existing index matches params.

That would mean that two scripts running at the same time could safely use the same index (they'd race on creation) and it would mean that if someone had a mismatch between index params and an existing index that would be detected and would error out.

Other option would be to run it at start before the ray part of the exercise.  To handle multiple scripts would still need the alternative API.